core/arch/riscv64/mangle.c

```diff
@@ -32,40 +32,170 @@
+
+    /* a0 is used to save and restore the pc and csr registers. */
+    cci->reg_skip[DR_REG_A0 - DR_REG_START_GPR] = false;
+
+    max_offs = get_clean_call_switch_stack_size();
+
+    PRE(ilist, instr,
+        INSTR_CREATE_addi(dcontext, opnd_create_reg(DR_REG_SP),
+                          opnd_create_reg(DR_REG_SP),
+                          opnd_add_flags(opnd_create_immed_int(-max_offs, OPSZ_12b),
+                                         DR_OPND_IMM_PRINT_DECIMAL)));
+
+    /* Push GPRs. */
+    for (int i = 1; i < DR_NUM_GPR_REGS; i += 1) {
+        if (cci->reg_skip[i])
+            continue;
+
+        PRE(ilist, instr,
+            INSTR_CREATE_sd(dcontext,
+                            opnd_add_flags(opnd_create_base_disp(
+                                               DR_REG_SP, DR_REG_NULL, 0,
+                                               dstack_offs + i * sizeof(reg_t), OPSZ_8),
+                                           DR_OPND_IMM_PRINT_DECIMAL),
+                            opnd_create_reg(DR_REG_ZERO + i)));
+    }
+
+    dstack_offs += DR_NUM_GPR_REGS * XSP_SZ;
+
+    if (opnd_is_immed_int(push_pc)) {
+        PRE(ilist, instr,
+            XINST_CREATE_load_int(dcontext, opnd_create_reg(DR_REG_A0), push_pc));
+    } else {
+        ASSERT(opnd_is_reg(push_pc));
+        reg_id_t push_pc_reg = opnd_get_reg(push_pc);
+        /* push_pc opnd is already pushed on the stack. */
+        PRE(ilist, instr,
+            INSTR_CREATE_ld(dcontext, opnd_create_reg(DR_REG_A0),
+                            OPND_CREATE_MEM64(DR_REG_SP, REG_OFFSET(push_pc_reg))));
+    }
+
+    PRE(ilist, instr,
+        INSTR_CREATE_sd(dcontext, OPND_CREATE_MEM64(DR_REG_SP, dstack_offs),
+                        opnd_create_reg(DR_REG_A0)));
+
+    dstack_offs += XSP_SZ;
+
+    /* Push FPRs. */
+    for (int i = 0; i < DR_NUM_FPR_REGS; i += 1) {
+        PRE(ilist, instr,
+            INSTR_CREATE_fsd(dcontext,
+                             opnd_add_flags(opnd_create_base_disp(
+                                                DR_REG_SP, DR_REG_NULL, 0,
+                                                dstack_offs + i * sizeof(reg_t), OPSZ_8),
+                                            DR_OPND_IMM_PRINT_DECIMAL),
+                             opnd_create_reg(DR_REG_F0 + i)));
+    }
+
+    dstack_offs += DR_NUM_FPR_REGS * XSP_SZ;
+
+    /* csrr a0, fcsr */
+    PRE(ilist, instr,
+        INSTR_CREATE_csrrs(dcontext, opnd_create_reg(DR_REG_A0),
+                           opnd_create_reg(DR_REG_X0),
+                           /* FIXME i#3544: Use register. */
+                           opnd_create_immed_int(0x003 /* fcsr */, OPSZ_12b)));
+
+    PRE(ilist, instr,
+        INSTR_CREATE_sd(dcontext, OPND_CREATE_MEM64(DR_REG_SP, dstack_offs),
+                        opnd_create_reg(DR_REG_A0)));
+
+    dstack_offs += XSP_SZ;
+
+    /* Keep mcontext shape. */
+    dstack_offs += (proc_num_simd_registers() * sizeof(dr_simd_t));
```

Comment:
Wait, is pushing the SIMD registers not yet implemented?  Please add a TODO comment for that so it's clear something is missing.